table/block_based/block_based_table_reader.h

```diff
@@ -755,7 +767,7 @@ class BlockBasedTableIterator : public InternalIteratorBase<TValue> {
   UserComparatorWrapper user_comparator_;
   InternalIteratorBase<IndexValue>* index_iter_;
   PinnedIteratorsManager* pinned_iters_mgr_;
-  TBlockIter block_iter_;
+  TBlockIter* block_iter_;
```

Comment:
Do we have to make the change? The main purpose of getting into this complicated template business is to get ride of this allocation. Now we are getting it back to a pointer, which is a little bit hard to swallow.

Reply:
> Do we have to make the change? The main purpose of getting into this complicated template business is to get ride of this allocation. Now we are getting it back to a pointer, which is a little bit hard to swallow.

This change was made to work around the problem that `TBlockIter` doesn't have move assignment support. A pool of `TBlockIter` is needed when implementing pre-loading blocks. The pool requires move assignment support, so I used `TBlockIter*` as a workaround.

I think this change can be avoided by making a pool of block contents, i.e. `CachableEntry<Block>`, because it already has move assignment support. `BlockBasedTable::RetrieveBlock` can be called to push `CachableEntry` into the pool in the loader loop. The entry can then be popped out in `BlockBasedTable::NewDataBlockIterator`. Please correct me if I missed something.

Reply:
I've eliminated the pointer change in the latest commits by pre-loading CachableEntry<Block> instead of DataBlockIterator.

Reply:
> I think the table builder part mostly looks good to me. I have questions on the reader part. If my understanding is correct, only one thread is used for decompression. Then what's the point of it? Do we plan to eventually make it multi-threaded?

As for the multi-threaded support for reader, since a major performance improvement might come from large scan tasks, I'm planning to make it a separate PR in the future after I've collected convincing experiment results for the performance it gains in such tasks.